proof-systems/src/darlin/mod.rs

```diff
@@ -0,0 +1,223 @@
+            .map_err(|_| FinalDarlinError::Other("Failed to convert usr ins to field elements".to_owned()))?;
+
+        Ok(FinalDarlinPCD::<G1, G2, D>::new(proof, usr_ins))
+    }
+
+    /// Fully verify a `FinalDarlinProof` from the PCDCircuit `C`, using the PCD implementation for 
+    /// the FinalDarlinPCD.
+    pub fn verify<R: RngCore>(
+        index_vk:     &FinalDarlinVerifierKey<G1::ScalarField, InnerProductArgPC<G1, D>>,
+        pc_vk_g1:     &DLogVerifierKey<G1>,
+        pc_vk_g2:     &DLogVerifierKey<G2>,
+        usr_ins:      &[G1::ScalarField],
+        proof:        &FinalDarlinProof<G1, G2, D>,
+        rng:          &mut R,
+    ) -> Result<bool, FinalDarlinError>
+    {
+        let final_darlin_pcd = FinalDarlinPCD::<G1, G2, D>::new(
+            proof.clone(), usr_ins.to_vec()
+        );
+
+        let final_darlin_pcd_vk = FinalDarlinPCDVerifierKey::<G1, G2, D>{
+            final_darlin_vk: index_vk,
+            dlog_vks: (pc_vk_g1, pc_vk_g2)
+        };
+
+        let res = final_darlin_pcd.verify(&final_darlin_pcd_vk, rng)?;
+
+        Ok(res)
+    }
+
+    /// Verifies only the IOP part of a `FinalDarlinProof`, i.e. a Marlin AHP 
+    /// for the PCDCircuit with correctly combined system and user inputs.
+    pub fn verify_ahp(
+        index_vk:       &FinalDarlinVerifierKey<G1::ScalarField, InnerProductArgPC<G1, D>>,
+        usr_ins:        &[G1::ScalarField],
+        proof:          &FinalDarlinProof<G1, G2, D>,
+    )  -> Result<(
+        QuerySet<'a, G1::ScalarField>,
+        Evaluations<'a, G1::ScalarField>,
+        Vec<LabeledCommitment<Commitment<G1>>>,
+        <InnerProductArgPC<G1, D> as PolynomialCommitment<G1::ScalarField>>::RandomOracle,
+    ), FinalDarlinError>
+    {
+        // Get "system inputs"
+        let mut public_inputs = proof.deferred.to_field_elements().unwrap();
+
+        // Append user inputs
+        public_inputs.extend_from_slice(usr_ins);
+
+        // Verify AHP
+        let res = Marlin::<G1::ScalarField, InnerProductArgPC<G1, D>, D>::verify_ahp(
+            index_vk, public_inputs.as_slice(), &proof.proof
+        )?;
+
+        Ok(res)
+    }
+
+    /// Verifies the dlog open part of a `FinalDarlinProof`. This also checks the 
+    /// "hard part" of the opening proof.
+    pub fn verify_opening(
+        pc_vk:          &DLogVerifierKey<G1>,
+        proof:          &FinalDarlinProof<G1, G2, D>,
+        labeled_comms:  Vec<LabeledCommitment<Commitment<G1>>>,
+        query_set:      QuerySet<'a, G1::ScalarField>,
+        evaluations:    Evaluations<'a, G1::ScalarField>,
+        fs_rng:         &mut <InnerProductArgPC<G1, D> as PolynomialCommitment<G1::ScalarField>>::RandomOracle,
+    ) -> Result<bool, FinalDarlinError>
+    {
+        let res = Marlin::<G1::ScalarField, InnerProductArgPC<G1, D>, D>::verify_opening(
+            pc_vk, &proof.proof, labeled_comms, query_set, evaluations, fs_rng
+        )?;
+
+        Ok(res)
+    }
```

Comment:
Why do we have an extra fn verify_opening for the dlog opening proof? Couldn't we just use the PCD implementation for FinalDarlinPCD?

Reply:
I thought to give it the same functions as in Marlin (as we decided with Algaro)